src/software/new_geom/segment.cpp

```diff
@@ -0,0 +1,137 @@
+#include "software/new_geom/segment.h"
+
+Segment::Segment(const Point& start, const Point& end) : start(start), end(end)
+{
+    if (start == end)
+    {
+        throw std::invalid_argument("Attempting to construct a degnerate segment");
+    }
+}
+
+void Segment::setStart(const Point& o)
+{
+    start = o;
+}
+
+const Point& Segment::getStart() const
+{
+    return start;
+}
+
+void Segment::setEnd(const Point& o)
+{
+    end = o;
+}
+
+const Point& Segment::getEnd() const
+{
+    return end;
+}
+
+Segment Segment::reverse() const
+{
+    return Segment(end, start);
+}
+
+Vector Segment::toVector() const
+{
+    return end - start;
+}
+
+double Segment::length() const
+{
+    return (end - start).length();
+}
+
+bool Segment::operator==(const Segment& other) const
+{
+    return start == other.start && end == other.end;
+}
+
+bool Segment::contains(const Point& point, double fixed_epsilon, int ulps_distance) const
```

Comment:
should this be in `contains.cpp`?

Reply:
talked to @jonathanlew offline, mb didn't realize there was another PR that is doing that